tensorflow/contrib/makefile/Makefile

```diff
@@ -373,12 +373,67 @@ $(MARCH_OPTION) \
 	ifdef ENABLE_EXPERIMENTAL_HEXNN_OPS
 		CXXFLAGS += -DENABLE_EXPERIMENTAL_HEXNN_OPS
 	endif
-	
-	OBJDIR := $(OBJDIR)android_$(ANDROID_ARCH)/
-	LIBDIR := $(LIBDIR)android_$(ANDROID_ARCH)/
-	BINDIR := $(BINDIR)android_$(ANDROID_ARCH)/
-	DEPDIR := $(DEPDIR)android_$(ANDROID_ARCH)/
 
+	ifeq ($(BUILD_FOR_TEGRA),1)
+		NVCC := $(JETPACK)/cuda-8.0/bin/nvcc
+		NVCCFLAGS := -x=cu -D__CUDACC__ -DNVCC -DNVIDIA_TEGRA -ccbin $(NDK_ROOT)/toolchains/$(TOOLCHAIN)/prebuilt/$(ANDROID_HOST_OS_ARCH)/bin/$(BIN_PREFIX)-g++ --std c++11 --expt-relaxed-constexpr -m64 -gencode arch=compute_53,\"code=sm_53\" -gencode arch=compute_62,\"code=sm_62\" -DEIGEN_AVOID_STL_ARRAY -DTENSORFLOW_USE_EIGEN_THREADPOOL -DLANG_CXX11 -DEIGEN_HAS_C99_MATH -DGOOGLE_CUDA=1 -DTF_EXTRA_CUDA_CAPABILITIES=5.3
+		CXXFLAGS4NVCC =\
+-DIS_SLIM_BUILD \
+-DNVIDIA_TEGRA \
+-fno-exceptions \
+-DNDEBUG $(OPTFLAGS) \
+-march=armv8-a \
+-fPIE \
+--sysroot $(NDK_ROOT)/platforms/android-21/arch-arm64
+
+		CXXFLAGS +=\
+-DGOOGLE_CUDA=1 \
+-DNVIDIA_TEGRA \
+-DEIGEN_AVOID_STL_ARRAY \
+-DEIGEN_HAS_C99_MATH \
+-DLANG_CXX11 -DTENSORFLOW_USE_EIGEN_THREADPOOL -DTF_EXTRA_CUDA_CAPABILITIES=5.3
+
+		INCLUDES += \
+-Itensorflow/core/kernels \
+-I$(JETPACK)/cuda-8.0/include \
+-I$(JETPACK) \
+-I$(JETPACK)/cuDNN/aarch64 \
+-I$(JETPACK)/cuda-8.0/extras/CUPTI/include \
+-I. \
+-Itegra \
+-Itegra/external/cub_archive/cub/device
```

Comment:
What is this for? Do we really need this whole archive?

Reply:
-I. \ can be removed.

-I$(JETPACK)/cuda-8.0/extras/CUPTI/include and libcupti.so is used by tensorflow/core/platform/default/gpu_tracer.cc. If we don’t need gpu_tracer.cc we can remove this.

+-Itegra \ contains cuda_config.h is written by google. I don't think it plays any practical role, but it is included in many CUDA source files.

+-Itegra/external/cub_archive/cub/device is a lot of cuda headers.
Most of them can be found in Jetpack, but the way TF code include headers makes it unable to use the copy in Jetpack
For example, 
tensorflow/core/kernels/depthwise_conv_op_gpu.cu.cc:25: #include "external/cub_archive/cub/util_ptx.cuh"
There is a util.ptx.cuh in Jetpack_Android_3.0/cuda-8.0/targets/aarch64-linux-androideabi/include/thrust/system/cuda/detail/cub/util_ptx.cuh
However, I can’t handle the path in the inclusion “external/cub_archive/”.

Also, I notice there are at least two cuh files that TF needs but don’t exists in Jetpack.
tegra/external2/cub_archive/cub/device/device_segmented_radix_sort.cuh
tegra/external2/cub_archive/cub/device/device_segmented_reduce.cuh.cuh

Reply:
Yes, let's disable GPU-tracer for now and worry about supporting it later if necessary.

Do we need to use tegra/cuda_config.h vs third_party/toolchains/gpus/cuda/cuda/cuda_config.h?

Is "Jetpack_Android_3.0/cuda-8.0/targets/aarch64-linux-androideabi/include/thrust/system/cuda/detail/cub" completely analogous to "external/cub_archive"? If so maybe we can create a symlink instead?

For device_segmented_radix_sort.cuh and device_segmented_reduce.cuh.cuh, what happens if we just don't compile the including files?

Reply:
The third_party/toolchains/gpus/cuda/cuda/cuda_config.h seems a little different, but it should be usable.

With a quick glance, the difference between Jetpack_Android_3.0/cuda-8.0/targets/aarch64-linux-androideabi/include/thrust/system/cuda/detail/cub and external/cub_archive are device_segmented_radix_sort.cuh and device_segmented_reduce.cuh.cuh

I will take a closer look on Monday.

If we don't include these two files. the compilation will fail. We need to exclude more than 10 source codes which seems unacceptable to me.

Reply:
I tested that I can use third_party/toolchains/gpus/cuda/cuda/cuda_config.h.
For cub_archive, I found that several file is missing in Jetpack_Android_3.0/cuda-8.0/targets/aarch64-linux-androideabi/include/thrust/system/cuda/detail/cub, such as
dispatch/dispatch_reduce.cuh
tegra/external2/cub_archive/cub/device/device_segmented_radix_sort.cuhtegra/external2/cub_archive/cub/device/device_segmented_reduce.cuh.cuh
dispatch_reduce_by_key.cuh 
dispatch_scan.cuh

In conclusion, I think we need to download cub archive during compilation.
I got it by using bazel to build for Java:
./configure #choose Y for CUDA
bazel build --config opt \
  //tensorflow/java:tensorflow \
  //tensorflow/java:libtensorflow_jni

Can you point to me the url of cub archive?

Reply:
For the bazel build the cub archive is downloaded via https://github.com/tensorflow/tensorflow/blob/master/tensorflow/workspace.bzl#L772

If it's necessary, we can do a similar download of the archive for Make builds in download_dependencies.sh

Reply:
Should be pretty simple to add it here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/makefile/download_dependencies.sh

Reply:
For CUPTI, I found another dependancy chain that prevents me to remove libcupti.so
tensorflow/core/common_runtime/direct_session.cc -> tensorflow/core/platform/default/gpu_tracer.cc -> tensorflow/core/platform/default/gpu/cupti_wrapper.cc -> libcupti.so

Without direct_session, benchmark fails with
session.cc:58 Not found: No session factory registered for the given session options: {target: "" config: } Registered factories are {}

Should we keep libcupti.so?

Reply:
Also, I noticed that TF is specifically looking for libcupti.so.8.0. It will throw an error if that lib doesn't exitst.
native : statusor.h:212 Non-OK-status: status_ status: Failed precondition: could not dlopen DSO: libcupti.so.8.0; dlerror: dlopen failed: library "libcupti.so.8.0" not found
I can't find where libcupti.so.8.0 is referenced in the source code.